src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs

```diff
@@ -270,11 +270,13 @@ private BoundExpression BindArgListOperator(InvocationExpressionSyntax node, Dia
                             break;
 
                         case BoundKind.ThisReference:
-                            if (InConstructorInitializer && receiver.WasCompilerGenerated)
+                            // Can't call the HasThis method due to EE doing odd things with containing member and its containing type.
+                            if ((InConstructorInitializer || InFieldInitializer) && receiver.WasCompilerGenerated)
```

Comment:
What happens if `this` is used in a static ctor / field initializer with dynamic?  Will this code path get hit?  If so do we have a test to make sure we don't crash?